UNIX/SOURCE/xwave-2.0/xwave.1
.TH XWAVE 1 "4 February 1991" "X Version 11"
.SH NAME
xwave \- A real-time perspective mesh demo for X
.SH SYNOPSIS
.B xwave 
[
.I -amp
A
]
[
.I -bound
bits
]
[
.I -cols
C
]
[
.I -dampfactor
B
]
[
.I -demo
nameOrNum
]
[
.I -edit
]
[
.I -force
funcname
]
[
.I -forcecol
N
]
[
.I -forcefull
]
[
.I -forcepoint
R C
]
[
.I -forcerow
N
]
[
.I -omega
M
]
[
.I -omega2
M
]
[
.I -plotvel
]
[
.I -printdemo
nameOrNum
]
[
.I -root
]
[
.I -rows
R
]
[
.I -springconst
K
]
[
.I -stereo
]
[
.I -version
]
[
.I -wave
funcname
]

.SH DESCRIPTION
.LP
Xwave plots a perspective projection of a plane grid oscillating in
three-space in real time.  You can select the size of the grid, a
forcing function to apply (and where to apply it), and what wave propagation
model to use.  You can display a plot of the velocity of the points in the
grid as well as the position.
Double buffering is
simulated using a background pixmap.
Other options are described below in the "OPTIONS" section.
.LP
While the wave is being displayed, you can pan and zoom, stop and start
updates, and make other changes.
This is described below in the "INTERACTIVE USE" section.

.SH OPTIONS
.TP
.B -amp A
Select the amplitude of the forcing function.
Typical numbers are between 20 and 100.
Default amplitude is 100.
.TP
.B -bound bits
Select which corners and edges of the grid are bound and which are free.
By default, all corners and edges are bound.
If you specify a single "0" or "1" as the argument to -bound, it applies to
all eight corners and edges.
Otherwise, each character in the string applies to one corner or edge: if
that character is "0", the corresponding corner or edge is free; else it is
bound.
The sequence of corners and edges in the string is: lrlc, lr, lrhc, hc,
hrhc, hr, hrlc, lc, where l is low, h is high, r is row, and c is column.
For example, -bound "00111000" means the high column end of the grid,
including those corners, are bound, and all other corners and edges are free.
.TP
.B -cols C
Specify how many columns are in the grid.
The minimum number of columns is 2.
Larger numbers give more detailed  wave displays,
but take longer to compute and thus do not update as quickly.
The default is 25.
.TP
.B -dampfactor B
Specify the damping factor used in the "wave" propagation model.
See the -wave switch for more details.
The default is 0.001.
.TP
.B -demo nameOrNum
Select a demonstration to run.
Each demo is equivalent to a specific set of switches (which you can see
with the -printdemo command line option).
For a quick tour of the program, run "xwave -demo N" starting with N=0 and
incrementing until it tells you there are no more demos.
.TP
.B -edit
Allows the user to interactively shape the wave while the program is running.
When the program is in edit mode, the mouse controls an X-shaped cursor that
moves over the surface of the grid.  Position the cursor to where you want
to modify the wave and press and hold the left mouse button.  At this point,
dragging the mouse up and down will modify the amplitude of the wave and
dragging it left and right will control the frequency.  Releasing the mouse
button will allow you to modify additional points.  When you are finished
editing points, press the right mouse button to enter continuous run mode.
.TP
.B -force funcname
Specify the name of the forcing function to use.
This can also be changed interactively (see the INTERACTIVE USE section).
.TP
.B -forcecol C
Specify that the selected forcing function is to be
applied to all grid points in the specified column.
(The multisine forcing function ignores this information.)
.TP
.B -forcefull
Specify that forcing functions which do not cover the full cycle should
force a zero value during the remainder of the cycle.
By default, the forcing function takes no action during this part of the
cycle, but lets the point move freely.
This affects the following forcing functions:
delta, pulse, onehalfsine, onehalfsinesquared.
.TP
.B -forcepoint R C
Specify that the selected forcing function is to be
applied to the specified grid point.
(The multisine forcing function ignores this information.)
.TP
.B -forcerow R
Specify that the selected forcing function is to be
applied to all grid points in the specified row.
(The multisine forcing function ignores this information.)
.TP
.B -omega M
Specify a frequency multiplier for forcing functions.
A larger number makes the forcing functions cycle through their values faster.
.TP
.B -omega2 M
Specify a secondary frequency multiplier.
This number is used in the following forcing functions:
pulse, onehalfsine, onehalfsinesquared.
If "-omega2 1" is specified, each of these functions will have a duty
cycle of 1/2; if "-omega 2" is specifed, it will be 1/4, etc.
.TP
.B -plotvel
Specify that a plot of the velocity of each point in the grid should appear
below the plot of the position of each point in the grid.
The velocity information is updated only by the "wave" propagation function.
.TP
.B -printdemo nameOrNum
Print the switches that comprise the specified demo, then exit.
.TP
.B -root
Plot the wave in the root window.
.TP
.B -rows R
Specify how many rows are in the grid.
The minimum number of rows is 2.
Larger numbers give more detailed  wave displays,
but take longer to compute and thus do not update as quickly.
The default is 18.
.TP
.B -springconst K
Specify the spring constant to use in the "wave" propagation function.
See the -wave switch for more details.
The default is 0.01.
.TP
.B -stereo
Plot separate projections in red and blue.  Make sure you have your
3-d glasses handy.  You also need a color display (obviously).
.TP
.B -version
Prints the current version of xwave and exits.
.TP
.B -wave funcname
Specify the name of the wave propagation function to use.
You can specify none, which allows you to see the "naked" forcing function.
You can specify relax, which gives a pleasant demo with the multisine
forcing function (the original xwave program).
You can specify wave, which uses a physical wave propagation model.
In the wave model, each point has a mass of 1 and a velocity as well as a
position.
A simple spring model is used, F = -K*x -B*v, where K is the spring constant
(specified with -springconst) and B is the damping factor (specified with
-dampfactor).
The distance used for x is the simple vertical distance between a point and
its neighbors.
Note that since this is a discreet simulation, the simulation is
highly dispersive (low frequencies travel faster),
so single pulses do not travel very well.
However, you can still see the effects of reflection against a free or
bound end, set up nice standing waves, see how the velocity is out of
phase with the position, etc.

.SH "INTERACTIVE USE"
.LP
While xwave is running, you can change the perspective by using the mouse
buttons.
You can pan the picture as it is running by holding down the middle mouse
button and moving the mouse.
You can zoom in and out by holding down the left or right mouse buttons.
(Note that these operations do not work unless xwave is running in
continuous run mode.)
.LP
There are a number of single-key keyboard commands you can enter into
the graphics window of xwave.
These commands can be entered whether or not xwave is in continuous run mode.
If the command is not understood, the bell is rung.
Note that there may be a bit of delay in processing a keyboard command when
the program is in continuous run mode.
.TP
.B c
Clear the data and reset the forcing function time; everything starts over
again.
.TP
.B d
Load the forcing function data.
This is useful when in step mode.
.TP
.B f
Select a different forcing function.
The new forcing function name is printed out.
.TP
.B p
Print out data.
Prints out all velocity and position data which is non-zero.
Mainly useful in step mode for debugging.
.TP
.B q
Quit the program.
.TP
.B r
Toggle run/step mode.
In run mode, the program runs continuously and updates the screen.
This consumes 100% of the available cpu time, and looks nice.
In step mode, you can step through each generation by pressing the
space bar, and see in detail how the grid changes at each step.
By default, the program comes up in continuous run mode.
.TP
.B t
Toggle timing mode.
When the program is in time mode, it prints an approximation of how many
frames per second it is displaying.  This is re-calculated and printed
every 50 frames.
.TP
.B w
Select a different wave propagation function.
The new propagation function name is printed out.
.TP
.B <space>
Step (when in step mode).
Calculates and displays the grid at the next time point.

.SH AUTHORS
.LP
Paul Riddle (paulr@umbc3.umbc.edu)
.br
Mike Friedman (mikef@umbc3.umbc.edu)
.br
Jim McBeath (jimmc@hisoft.uucp)



UNIX/SOURCE/xwave-2.0/xwave.c
/******************************************************************************
** xwave - A 3-d perspective mesh for X
**
** Original Authors: (Jan 1991)
** Mike Friedman (mikef@umbc3.umbc.edu)
** Paul Riddle (paulr@umbc3.umbc.edu)
**
** Additional Help:
** Joe  Poole  (joe@umbc3.umbc.edu)
**
** University of Maryland, Baltimore County
*
* Enhanced by Jim McBeath (jimmc@hisoft.uucp) Feb 1991
******************************************************************************/

#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xos.h>
#include <stdio.h>
#include <math.h>
#include "xwave.h"
#include "patchlevel.h"
#include <sys/types.h>
#include <sys/time.h>


#define   DFLTMAXROWS    18
#define   DFLTMAXCOLS    25
#define   GENS       2
#define   SINTABSIZE    400
#define   TWOPI      6.283

#define   NFRAMES    50

#define   ROOT       0x1

#define SPRINGCONST 0.01
#define DAMPFACTOR 0.001

extern double atof();

extern int (*ForceFunc)();
extern int (*WaveFunc)();

extern Display *dpy;
extern int ampl;

extern int numEditedPoints;

int maxrows = DFLTMAXROWS;
int maxcols = DFLTMAXCOLS;
int CurIndex = 0;
int NextIndex = 1;
struct pinfo *Generation[GENS];
struct pinfo NullPinfo = {0};
struct pinfo *CurGen;
struct pinfo *NextGen;

int *amplitudes;
int *frequencies;
int *editedRowVals;
int *editedColVals;

double sinTab[SINTABSIZE];
int sinTabSize = SINTABSIZE;

int running = 1;
int fftime;
int omega=5;
int omega2=10;
int forcerow = -1;
int forcecol = -1;
int forceptr = -1;
int forceptc = -1;
int forcefull = 0;

double springconst = SPRINGCONST;
double dampfactor = DAMPFACTOR;

int flagtime=0;
int flagstereo=0;
int flagroot=0;
int flagplotvel=0;
int flagedit=0;
int didedit=0;

int nf;
int start;


char *
smalloc(s)
int s;
{
	extern char *malloc();
	char *p;

	p = malloc(s);
	if (p) return p;
	fprintf(stderr,"No more memory\n");
	exit(2);
}

char *
scalloc(n,s)
int n,s;
{
	extern char *calloc();
	char *p;

	p = calloc(n,s);
	if (p) return p;
	fprintf(stderr,"No more memory\n");
	exit(2);
}

char *demoorig[] = { "original" };	/* the original - no args */
char *demodeltapoint[] = { "deltapoint", "-force", "delta", "-wave", "wave",
	"-omega", "2" };
char *demodeltarow[] = { "deltarow", "-force", "delta", "-wave", "wave",
	"-forcerow", "8", "-amp", "30", "-omega", "2", "-bound", "0" };
char *demosineh[] = { "sineh", "-force", "onehalfsinesquared", "-wave", "wave",
	"-rows", "2", "-cols", "50", "-forcecol", "0", "-bound", "00111000",
	"-omega", "1", "-omega2", "4" };
char *demosine[] = { "sine", "-force", "sine", "-wave", "wave",
	"-rows", "2", "-cols", "48", "-forcecol", "0", "-bound", "00111000",
	"-omega", "5", "-plotvel", "-springconst", "0.05" };
char *demopulse[] = { "pulse", "-force", "pulse", "-forcefull", "-wave", "wave",
	"-omega", "1", "-omega2", "4", "-rows", "2", "-cols", "46",
	"-forcecol", "0", "-bound", "00111000", "-springconst", "0.05",
	"-plotvel" };
char *demopulse2[] = { "pulse", "-force", "pulse", "-wave", "wave",
	"-omega", "2", "-omega2", "4" };
/* remember to add new demo to demoinfo table below */

struct {
	char **p;
	int n;
} demoinfo [] = {
	{ demoorig, sizeof(demoorig)/sizeof(demoorig[0]) },
	{ demodeltapoint, sizeof(demodeltapoint)/sizeof(demodeltapoint[0]) },
	{ demodeltarow, sizeof(demodeltarow)/sizeof(demodeltarow[0]) },
	{ demosineh, sizeof(demosineh)/sizeof(demosineh[0]) },
	{ demosine, sizeof(demosine)/sizeof(demosine[0]) },
	{ demopulse, sizeof(demopulse)/sizeof(demopulse[0]) },
	{ demopulse2, sizeof(demopulse2)/sizeof(demopulse2[0]) },
};
int demoinfocount = sizeof(demoinfo)/sizeof(demoinfo[0]);

int
findDemo(s)
char *s;
{
	int i;

	i = atoi(s);
	if (i>0 || strcmp(s,"0")==0) {
		if (i>=demoinfocount) {
			(void)fprintf(stderr,"Max demo number is %d\n",
				demoinfocount-1);
			exit(1);
		}
		return i;	/* he specified a number (index) */
	}
	for (i=0; i<demoinfocount; i++) {
		if (strcmp(demoinfo[i].p[0],s)==0)
			return i;
	}
	(void)fprintf(stderr,"No such demo switch %s\n",s);
	exit(1);
}

DemoSwitch(s)
char *s;
{
	int n;

	n = findDemo(s);
	ProcessArgs(demoinfo[n].n,demoinfo[n].p);
}

PrintDemoSwitch(s)
char *s;
{
	int i,n;

	n = findDemo(s);
	for (i=0; i<demoinfo[n].n; i++) {
		printf("%s ",demoinfo[n].p[i]);
	}
	printf("\n");
	exit(0);
}

ListDemoSwitches(fp)
FILE *fp;
{
	int i;

	fprintf(fp,"Demo switches are:\n");
	for (i=0; i<demoinfocount; i++) {
		fprintf(fp," %s", demoinfo[i].p[0]);
	}
	fprintf(fp,"\n");
}

ProcessArgs(argc,argv)
int argc;
char *argv[];
{
	int i;

	for (i=1; i<argc; i++) {
                if (strcmp (argv[i], "-edit")==0) {
                        flagedit = 1;
                        didedit = 1;
		}
		else if (strcmp (argv[i], "-root")==0)
			flagroot = 1;
		else if (strcmp (argv[i], "-stereo")==0)
			flagstereo = 1;
		else if (strcmp(argv[i],"-springconst")==0) {
			springconst = atof(argv[++i]);
		}
		else if (strcmp(argv[i],"-dampfactor")==0) {
			dampfactor = atof(argv[++i]);
		}
		else if (strcmp(argv[i],"-rows")==0) {
			maxrows = atoi(argv[++i]);
			if (maxrows<2) maxrows = 2;
                        else if (maxrows>25) maxrows = 25;
		}
		else if (strcmp(argv[i],"-cols")==0) {
			maxcols = atoi(argv[++i]);
			if (maxcols<2) maxcols = 2;
		}
		else if (strcmp(argv[i],"-omega")==0) {
			omega = atoi(argv[++i]);
			if (omega<=0) omega=1;	/* can't divide by 0 */
		}
		else if (strcmp(argv[i],"-omega2")==0) {
			omega2 = atoi(argv[++i]);
			if (omega2<=0) omega2=1;	/* can't divide by 0 */
		}
		else if (strcmp(argv[i],"-amp")==0) {	/* forcing amplitude */
			ampl = atoi(argv[++i]);
			if (ampl<=0) ampl = 100;
		}
		else if (strcmp(argv[i],"-force")==0) {
			if (!SetNamedForceFunc(argv[++i])) {
				usage();
				exit(1);
			}
		}
		else if (strcmp(argv[i],"-forcerow")==0) {
			forcerow = atoi(argv[++i]);
		}
		else if (strcmp(argv[i],"-forcecol")==0) {
			forcecol = atoi(argv[++i]);
		}
		else if (strcmp(argv[i],"-forcepoint")==0) {
			forceptr = atoi(argv[++i]);
			forceptc = atoi(argv[++i]);
		}
		else if (strcmp(argv[i],"-forcefull")==0) {
			forcefull = 1;
		}
		else if (strcmp(argv[i],"-bound")==0) {
			SetBound(argv[++i]);
		}
		else if (strcmp(argv[i],"-wave")==0) {
			if (!SetNamedWaveFunc(argv[++i])) {
				usage();
				exit(1);
			}
		}
		else if (strcmp(argv[i],"-demo")==0) {
			DemoSwitch(argv[++i]);
		}
		else if (strcmp(argv[i],"-printdemo")==0) {
			PrintDemoSwitch(argv[++i]);
		}
		else if (strcmp(argv[i],"-plotvel")==0) {
			flagplotvel = 1;
		}
		else if (strcmp(argv[i],"-version")==0) {
			(void)printf("xwave version %d.%d, %s\n",
				VERSION, PATCHLEVEL, VDATE);
			exit(0);
		}
		else {
			usage();
			exit (1);
		}
	}
}

CheckArgs()
{
	if (forcerow >= maxrows)
		forcerow = maxrows/2;
	if (forcecol >= maxcols)
		forcecol = maxcols/2;
	if (forceptr >= maxrows)
		forceptr = maxrows/2;
	if (forceptc >= maxcols)
		forceptc = maxcols/2;
	if (forcerow<0 && forcecol<0) {
		if (forceptr<0)
			forceptr = maxrows/2;
		if (forceptc<0)
			forceptc = maxcols/2;
	} else {
		forceptr = forceptc = -1;
	}
}

usage()
{
	(void)fprintf(stderr, "Usage: xwave [-edit] [-root] [-stereo] [-springconst K]\n");
	(void)fprintf(stderr, "   [-dampfactor B] [-rows r] [-cols c]\n");
	(void)fprintf(stderr, "   [-force func] [-wave func] [-bound bits]\n");
	(void)fprintf(stderr, "   [-omega n] [-omega2 n] [-amp n]\n");
	(void)fprintf(stderr, "   [-forcerow r] [-forcecol c] [-forcepoint r c]\n");
	(void)fprintf(stderr, "   [-forcefull]\n");
	(void)fprintf(stderr, "   [-demo D] [-printdemo D] [-plotvel] [-version]\n");
	PrintForceFuncs(stderr);
	PrintWaveFuncs(stderr);
	ListDemoSwitches(stderr);
}

AllocData()
{
	Generation[0] = (struct pinfo *)smalloc(sizeof(struct pinfo)*
					maxrows*maxcols);
	Generation[1] = (struct pinfo *)smalloc(sizeof(struct pinfo)*
					maxrows*maxcols);
        amplitudes    = (int *)scalloc(maxrows*maxcols,sizeof(int));
        frequencies   = (int *)scalloc(maxrows*maxcols,sizeof(int));
        editedRowVals = (int *)scalloc(maxrows,sizeof(int));
        editedColVals = (int *)scalloc(maxcols,sizeof(int));
}

ClearData()
{
	int r, c, g;

	/* clear the forcing function timer */
	fftime = 0;

	/* Zero out the generation matrix */
	for (r = 0; r < maxrows; r++)
		for (c = 0; c < maxcols; c++)
			for (g = 0; g<GENS; g++)
				Generation[g][IX(r,c)] = NullPinfo;

	CurIndex = 0;
	NextIndex = 1;
	CurGen = Generation[CurIndex];
	NextGen = Generation[NextIndex];
}

/* This routine initializes the sine table.  The table is accessed during
 * computation to speed things up somewhat. */
void setupSinTable (sinTab)
double *sinTab;
{
	int i;
	double t = 0.0;

	for (i = 0; i < SINTABSIZE; i++, t += TWOPI / SINTABSIZE)
		sinTab[i] = sin (t);
}

DumpCur()
{
	int r,c;
	int v,p;

	for (r = 0; r < maxrows; r++) {
		for (c = 0; c < maxcols; c++) {
			v = CurGen[IX(r,c)].vel;
			p = CurGen[IX(r,c)].pos;
			if (v!=0 || p!=0)
				printf("[%d,%d]=(%d,%d)\n",r,c,v,p);
		}
	}
}

ProcessKey(s)
char *s;
{
	if (strcmp(s,"t")==0) {		/* time # frames */
                nf=0;
                flagtime=!flagtime;
	}
	else if (strcmp(s,"c")==0) {	/* clear */
		ClearData();
		ShowData();
	}
	else if (strcmp(s,"d")==0) {	/* (load) data */
		(*ForceFunc)();
		ShowData();
	}
	else if (strcmp(s,"f")==0) {	/* forcing function */
		SetNextForceFunc();
	}
	else if (strcmp(s,"p")==0) {	/* print data */
		DumpCur();
	}
	else if (strcmp(s," ")==0) {	/* step */
		Step();
		ShowData();
	}
	else if (strcmp(s,"r")==0)	/* run mode */
		running = !running;
	else if (strcmp(s,"q")==0)	/* q to exit */
		exit(0);
	else if (strcmp(s,"w")==0) {	/* wave propagation function */
		SetNextWaveFunc();
	}
	else XBell(dpy,25);	/* unknown */
}

ProcessEvent()
{
	/*
	 * Handle resizes.  When a resize happens, we need to re-allocate
	 * the background pixmap.  Note that the bigger the window, the
	 * larger the area that needs to get copied, and therefore the
	 * slower the program gets.
	 */
	XEvent event;
	int t;
	char cbuf[20];

	XNextEvent (dpy, &event);
	switch (event.type) {
	case ConfigureNotify:
		HandleResize(&event);
		break;
	case KeyPress:
                if (!flagedit) {
		    t = XLookupString(&event,cbuf,sizeof(cbuf),
			(KeySym *)NULL, (XComposeStatus *)NULL);
		    if (t<=0) break;
                    flagedit = 0;
                    cbuf[t] = '\0';
		    ProcessKey(cbuf);
		    break;
	        }
	default:
		break;
	}
}

Step()
{
	(*WaveFunc)();	/* calculate values at the next time point */
	CurIndex ^= 1;	/* swap next and current buffers */
	NextIndex ^= 1;
	CurGen = Generation[CurIndex];
	NextGen = Generation[NextIndex];

	(*ForceFunc)();	/* apply forcing function */

}

main (argc, argv)
int argc;
char *argv[];
{
	int t;

	InitWaveFunc();
	InitForceFunc();

	ProcessArgs(argc,argv);
	CheckArgs();	/* check for consistency etc. */
	AllocData();
	ClearData();
	InitWindow ();
	setupSinTable (sinTab);
	if (flagroot)
		WaitForExpose();
	while (1) {
		if (running) {
                        if (flagtime) {
                               if (nf==0)
                                     start = time((time_t)0);
                               if (nf==NFRAMES) {
                                     printf("#frames/sec.~%f\n",
                                            (float)nf/((time((time_t)0)-start)));
                                     nf = 0;
                               }
                               else
                                     nf++;
		        }
			if (QLength (dpy)) {
				ProcessEvent();
			}
			DoQueryPointer();
                        if (flagedit)
                                edit();
			Step();
			ShowData();
		}
		else {
		        ProcessEvent();
                        if (flagedit)
                                edit();
		}
	}
	/* NOTREACHED */
}

/* end */







